File: backend/src/main/java/com/noura/platform/commerce/api/v1/dto/common/ApiEnvelope.java
package com.noura.platform.commerce.api.v1.dto.common;

import java.time.Instant;
import java.util.List;

public record ApiEnvelope<T>(
        boolean success,
        String code,
        String message,
        T data,
        List<String> details,
        Instant timestamp,
        String traceId
) {
    public static <T> ApiEnvelope<T> success(String code, String message, T data, String traceId) {
        return new ApiEnvelope<>(true, code, message, data, List.of(), Instant.now(), traceId);
    }

    public static ApiEnvelope<Void> error(String code, String message, List<String> details, String traceId) {
        List<String> safeDetails = details == null ? List.of() : details;
        return new ApiEnvelope<>(false, code, message, null, safeDetails, Instant.now(), traceId);
    }
}

File: archive/legacy-monolith/mobile-app/lib/core/network/models/api_envelope.dart
class ApiEnvelope {
  const ApiEnvelope({
    required this.success,
    this.message,
    this.data,
    this.error,
    this.timestamp,
    this.path,
  });

  factory ApiEnvelope.fromJson(Map<String, dynamic> json) {
    return ApiEnvelope(
      success: json['success'] == true,
      message: json['message'] as String?,
      data: json['data'],
      error: json['error'] is Map<String, dynamic>
          ? ApiErrorBody.fromJson(json['error'] as Map<String, dynamic>)
          : null,
      timestamp: json['timestamp'] as String?,
      path: json['path'] as String?,
    );
  }

  final bool success;
  final String? message;
  final Object? data;
  final ApiErrorBody? error;
  final String? timestamp;
  final String? path;
}

class ApiErrorBody {
  const ApiErrorBody({this.code, this.detail});

  factory ApiErrorBody.fromJson(Map<String, dynamic> json) {
    return ApiErrorBody(
      code: json['code'] as String?,
      detail: json['detail'] as String?,
    );
  }

  final String? code;
  final String? detail;
}
